### PR TITLE
python: Upgrade PyPerf

### PR DESCRIPTION
**Pull request by Jongy**
## Description
Upgrade PyPerf to include latest improvements (https://github.com/Granulate/bcc/pull/5,  https://github.com/Granulate/bcc/pull/7, https://github.com/Granulate/bcc/pull/8, https://github.com/Granulate/bcc/pull/9, https://github.com/Granulate/bcc/pull/3, https://github.com/Granulate/bcc/pull/4, https://github.com/Granulate/bcc/pull/10/files)

This includes more accurate error codes, smarter filename -> module path translation, and kernel stacks collection.

I also unified parts of the build (between `Dockerfile` & `pyi.Dockerfile`) to avoid duplication.

## Motivation and Context
Improve PyPerf :)

## How Has This Been Tested?
PyPerf features were tested separately.

gProfiler was tested with our tests.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:
- [x] My code follows the code style of this project.
- [ ] I have updated the relevant documentation.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Jongy** [comment]
Just gotta merge https://github.com/Granulate/bcc/pull/5 and I'll upgrade the PyPerf version here

**Comment by Jongy** [comment]
@d3dave do we want to update the events buffer size / symbols map size here as well? (as in, extra parameters passed on invocation to `PyPerf`). I think we can go ahead and increase the events buffer size even more.

**Comment by d3dave** [comment]
> I think we can go ahead and increase the events buffer size even more.

Then let's do it 👍

File: gprofiler/python.py [hunk 1]
```diff
@@ -120,6 +120,9 @@ def snapshot(self) -> Mapping[int, Mapping[str, int]]:
 
 class PythonEbpfProfiler(PythonProfilerBase):
     PYPERF_RESOURCE = "python/pyperf/PyPerf"
+    PYPERF_EVENTS_BUFFER_PAGES = 256  # 1mb and needs to be physically contiguous
+    # 28mb (each symbol is 224 bytes), but needn't be physicall contiguous so don't care
+    PYPERF_SYMBOLS_MAP_SIZE = 131072
```
Comment (d3dave) [inline gprofiler/python.py:125]:
```suggestion
    symbols_map_size = 131072
```

File: gprofiler/python.py [hunk 2]
```diff
@@ -120,6 +120,9 @@ def snapshot(self) -> Mapping[int, Mapping[str, int]]:
 
 class PythonEbpfProfiler(PythonProfilerBase):
     PYPERF_RESOURCE = "python/pyperf/PyPerf"
+    PYPERF_EVENTS_BUFFER_PAGES = 256  # 1mb and needs to be physically contiguous
```
Comment (d3dave) [inline gprofiler/python.py:123]:
```suggestion
    events_buffer_pages = 256  # 1mb and needs to be physically contiguous
```

**Review by d3dave — CHANGES_REQUESTED** [review]

**Comment by Jongy** [comment]
Pfft it's racy due to sometimes not catching kernel stacks :/

File: gprofiler/python.py (hunk 1 above)
Comment (Jongy) [reply gprofiler/python.py:125]: Why lowercase tho? consts, after all

Comment (d3dave) [reply gprofiler/python.py:125]: cosmetic

Comment (Jongy) [reply gprofiler/python.py:125]: Don't care enough. k

**Comment by Jongy** [comment]
Fixed the racy tests, I hope

**Review by d3dave — DISMISSED** [review]

**Comment by Jongy** [comment]
@d3dave please reapprove if it passes

**Comment by Jongy** [comment]
https://github.com/Granulate/bcc/pull/10/files as well

File: gprofiler/python.py [hunk 3]
```diff
@@ -256,6 +256,12 @@ def snapshot(self) -> Mapping[int, Mapping[str, int]]:
         collapsed_path = self._dump()
         collapsed_text = collapsed_path.read_text()
         collapsed_path.unlink()
+        # PyPerf outputs sampling & error counters every interval, print them.
+        # also, makes sure its output pipe doesn't fill up.
+        # using read1() which performs just a single read() call and doesn't read until EOF
+        # (unliked Popen.communicate())
```
Comment (d3dave) [inline gprofiler/python.py:262]:
```suggestion
        # (unlike Popen.communicate())
```

File: gprofiler/python.py [hunk 4]
```diff
@@ -256,6 +256,12 @@ def snapshot(self) -> Mapping[int, Mapping[str, int]]:
...
+        assert self.process is not None
+        logger.debug(f"PyPerf output: {self.process.stderr.read1()}")
```
Comment (d3dave) [inline gprofiler/python.py:264]: This should be in `_dump` in case an exception is thrown between it and this line.

**Review by d3dave — CHANGES_REQUESTED** [review]

File: gprofiler/python.py (hunk 4 above)
Comment (Jongy) [reply gprofiler/python.py:264]: Where can an exception be thrown? The only thing between is reading a file that we know exists, and unlinking it

Comment (Jongy) [reply gprofiler/python.py:264]: The read show be performed after the file is written because we want to see the latest output

Comment (d3dave) [reply gprofiler/python.py:264]: After `_wait_for_output_file` the file is written.

Comment (Jongy) [reply gprofiler/python.py:264]: IK, so between `_wait_for_output_file` to this current point, which exception can be raised?

> The only thing between is reading a file that we know exists, and unlinking it

Comment (d3dave) [reply gprofiler/python.py]: Eh, best to have the output logged asap

Comment (Jongy) [reply gprofiler/python.py]: Well since it's just the read of a file between, I don't think it matters much. But okay, will move

**Review by d3dave — APPROVED** [review]

**Pull request merged**